### PR TITLE
oracle-function: implement the function `to_date`

### PR DESCRIPTION
**Pull request by joccau**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/57815

Problem Summary:

### What changed and how does it work?
- Simple implementation of Oracle function ‘to_date’.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

### Manul test
```
mysql> select to_date('2024-11-23','yyyy-mm-dd');
+------------------------------------+
| to_date('2024-11-23','yyyy-mm-dd') |
+------------------------------------+
| 2024-11-23 00:00:00                |
+------------------------------------+

mysql> select to_date('2024-11-23','YYyY-Mm-Dd');
+------------------------------------+
| to_date('2024-11-23','YYyY-Mm-Dd') |
+------------------------------------+
| 2024-11-23 00:00:00                |
+------------------------------------+

mysql> select to_date('20241123','yyyymmdd');
+--------------------------------+
| to_date('20241123','yyyymmdd') |
+--------------------------------+
| 2024-11-23 00:00:00            |
+--------------------------------+

mysql> select to_date('2024-11-23 07:45:37','yyYy-Mm-dD hH24:Mi:Ss');
+--------------------------------------------------------+
| to_date('2024-11-23 07:45:37','yyYy-Mm-dD hH24:Mi:Ss') |
+--------------------------------------------------------+
| 2024-11-23 07:45:37                                    |
+--------------------------------------------------------+

mysql> select to_date('2024/11/23 07:45:37','yyYy/Mm/dD hH24:Mi:Ss');
+--------------------------------------------------------+
| to_date('2024/11/23 07:45:37','yyYy/Mm/dD hH24:Mi:Ss') |
+--------------------------------------------------------+
| 2024-11-23 07:45:37                                    |
+--------------------------------------------------------+

format is not right.
mysql> select to_date('2024-11-23 07:45:37','yyYy-Mm-dD hH24:Mi');
+-----------------------------------------------------+
| to_date('2024-11-23 07:45:37','yyYy-Mm-dD hH24:Mi') |
+-----------------------------------------------------+
| NULL                                                |
+-----------------------------------------------------+
1 row in set, 1 warning (0.00 sec)
```

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/executor/reload_expr_pushdown_blacklist.go [hunk 1]
```diff
@@ -172,6 +172,7 @@ var funcName2Alias = map[string]string{
 	"date":                       ast.Date,
 	"date_add":                   ast.DateAdd,
 	"date_format":                ast.DateFormat,
+	"to_date":                    ast.ToDate,
```
Comment (joechenrh) [inline pkg/executor/reload_expr_pushdown_blacklist.go:175]: Can we put all oracle function names together which may be more clearer? Like
```
// Oracle functions
"to_date": ast.ToDate,
"to_char": ast.ToChar,
.... 
```

File: pkg/parser/ast/functions.go [hunk 2]
```diff
@@ -157,6 +157,7 @@ const (
 	SecToTime        = "sec_to_time"
 	Second           = "second"
 	StrToDate        = "str_to_date"
+	ToDate           = "to_date"
```
Comment (joechenrh) [inline pkg/parser/ast/functions.go:160]: Ditto

Comment (joccau) [reply pkg/parser/ast/functions.go:160]: Good idea~.

File: pkg/expression/builtin.go [hunk 3]
```diff
@@ -725,6 +725,7 @@ var funcs = map[string]functionClass{
 	ast.SecToTime:        &secToTimeFunctionClass{baseFunctionClass{ast.SecToTime, 1, 1}},
 	ast.Second:           &secondFunctionClass{baseFunctionClass{ast.Second, 1, 1}},
 	ast.StrToDate:        &strToDateFunctionClass{baseFunctionClass{ast.StrToDate, 2, 2}},
+	ast.ToDate:           &toDateFunctionClass{baseFunctionClass{ast.ToDate, 1, 2}},
```
Comment (River2000i) [inline pkg/expression/builtin.go:728]: How about also be kept consistent with what @joechenrh mentioned when we create new functions in `expression` package🤔

Comment (joccau) [reply pkg/expression/builtin.go]: > How about also be kept consistent with what @joechenrh mentioned when we create new functions in `expression` package🤔

Maybe we could put ast.XXX(function name) together, but functionClass has own place.

**Review by joechenrh — APPROVED** [review]

File: pkg/expression/builtin_time.go [hunk 4]
```diff
@@ -1844,6 +1845,39 @@ func (b *builtinGetFormatSig) evalString(ctx EvalContext, row chunk.Row) (string
...
+		return nil, err
+	}
+	bf.setDecimalAndFlenForDatetime(types.MinFsp)
+	sig := &builtinStrToDateDatetimeSig{bf}
+	sig.setPbCode(tipb.ScalarFuncSig_StrToDateDatetime)
+	return sig, nil
+}
+
+func convertFormatFromOracelToMysql(format string) string {
+	re := regexp.MustCompile(`(?i)YYYY`)
```
Comment (River2000i) [inline pkg/expression/builtin_time.go:1868]: how about add some illegal pattern test?

Comment (joccau) [reply pkg/expression/builtin_time.go:1868]: > how about add some illegal pattern test?

Done

**Review by River2000i — APPROVED** [review]
lgtm

**Pull request merged**